Add Bazel Test Optimization setup docs (#37551)
DataDog/documentation: content/en/tests/setup/bazel/_index.md

---
title: Configure Test Optimization with Bazel
description: Configure Test Optimization for Bazel test targets with Datadog's official Bazel rules.
further_reading:
    - link: "/tests/setup/"
      tag: "Documentation"
      text: "Configure Test Optimization"
    - link: "/tests/explorer/"
      tag: "Documentation"
      text: "Explore Test Results and Performance"
    - link: "/tests/troubleshooting/"
      tag: "Documentation"
      text: "Troubleshooting Test Optimization"
---

> <div class="alert alert-info">Bazel support for Test Optimization is in Preview.</div>

Reviewer: Why do we say it's in preview?

In Datadog lingo, that generally means customers need to request access for the feature first: https://datadoghq.atlassian.net/wiki/spaces/PM/pages/4371088276/How+to+See+Preview+Requests+and+Enablement+Status#What%E2%80%99s-a-Preview%3F

Reviewer 2: if it's indeed in Preview, let's use the callout element:
```suggestion
{{< callout url="#" btn_hidden="true" header="false" >}}
Bazel support for Test Optimization is in Preview.
{{< /callout >}}
```
and if it has a page within https://www.datadoghq.com/product-preview/? we should include that link!

Author: Preview section was removed.

> Datadog provides official Bazel rules for Test Optimization. Use these rules to configure Bazel test targets. The rules read Test Optimization metadata, write local payload files during test execution, and upload test results to Datadog after Bazel tests complete.

Reviewer 2: I think it's nice to link to the public repo where they can see/explore everything, but I am curious what you think!

```suggestion
Datadog provides official Bazel rules for Test Optimization, which you can use to configure Bazel test targets. The rules read Test Optimization metadata, write local payload files during test execution, and upload test results to Datadog after Bazel tests complete. See all rules in the [`DataDog/rules_test_optimization` repository](https://github.com/DataDog/rules_test_optimization).
```

The Bazel integration keeps Datadog metadata fetches outside test execution. During module or repository resolution, Bazel fetches Test Optimization metadata from Datadog and exposes it through a generated repository. During test execution, language-specific macros pass the metadata location to the test process and configure payloads to be written under `TEST_UNDECLARED_OUTPUTS_DIR`. After tests finish, run the doctor and uploader targets with `bazel run`.

> Test Impact Analysis is not supported for Bazel.

Reviewer 2: ```suggestion
[Test Impact Analysis][1] is not supported for Bazel.
```

## Language setup pages

Use the language-specific setup page for your Bazel test targets:

- [Java tests][1]
- [Python tests][2]

> - [Go tests][3]

Reviewer 2: I like the idea of using logo cards here, both for consistency (with the parent page) and visual appeal
```suggestion

{{< card-grid card_width="75px" >}}
  {{< image-card href="/tests/setup/bazel/java/" src="integrations_logos/java_avatar.svg" alt="Java" >}}
  {{< image-card href="/tests/setup/bazel/python/" src="integrations_logos/python_avatar.svg" alt="Python" >}}
  {{< image-card href="/tests/setup/bazel/go/" src="integrations_logos/golang-avatar.png" alt="Go" >}}
{{< /card-grid >}}
```

## Compatibility

This section includes setup pages for the following language test targets:

| Language | Bazel macro | Notes |
|---|---|---|
| Java | `dd_topt_java_test` | Requires a `dd-java-agent` JAR label. |
| Python | `dd_topt_py_test` | Supports the managed `pytest` runner and repository-owned pytest wrappers. |

> | Go | `dd_topt_go_test` | Use `test_optimization` mode for the faster standard-library `testing` path, or `general` mode for broader Orchestrion support. |

Reviewer 2: I think the "Notes" value for each of these should link to the Compatibility sections on each of their respective pages

Use `datadog-rules-test-optimization` version `1.2.0` and the commit pin shown on each language setup page.

> ## How the Bazel flow works

Reviewer 2: ```suggestion
## How the Bazel setup flow works
```

> The Bazel setup has four parts:

Reviewer 2: ```suggestion
For detailed setup guides, see the [language-specific setup pages](#language-setup-pages). On a high level, the Bazel setup has four parts:
```

> 1. Add the `datadog-rules-test-optimization` module and the companion module for your language.

Reviewer 2: feels weird to say "add" but not say where
```suggestion
1. Add the `datadog-rules-test-optimization` module and the companion module for your language to `MODULE.bazel`.
```

1. Configure a sync repository with your Datadog service name, runtime name, and runtime version.
1. Replace the language test rule with the Datadog Bazel macro for each instrumented test target.
1. Run tests, validate local payloads with the doctor target, validate enrichment with the uploader dry run, and then upload payloads.

For remote execution, configure Bazel to download test outputs locally:

```text
test:test-optimization --remote_download_outputs=all
```

Without local `test.outputs` directories, the doctor and uploader cannot inspect payload files after `bazel test`.

## Upload payloads

Run the upload flow after your Bazel tests complete. Replace `//tools/test_optimization` with the package where you declared the doctor and uploader targets:

```bash
bazel test --config=test-optimization //...
bazel run --config=test-optimization //tools/test_optimization:dd_test_optimization_doctor
bazel run --config=test-optimization //tools/test_optimization:dd_upload_payloads -- --dry-run --validate-enrichment

> DD_API_KEY=<DATADOG_API_KEY> DD_SITE=<DATADOG_SITE> bazel run --config=test-optimization //tools/test_optimization:dd_upload_payloads

Author: this is an example snipped not a script.

```

Do not pass `DD_API_KEY`, `DD_SITE`, `DD_GIT_*`, or upload endpoint variables through `--test_env`. Forward sync metadata with `--repo_env`, and pass upload credentials only to the uploader runtime.

## Further reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /tests/setup/bazel/java/
[2]: /tests/setup/bazel/python/

> [3]: /tests/setup/bazel/go/

Reviewer 2: ```suggestion
[1]: /tests/test_impact_analysis/
```

further_reading:
    - link: "/tests/setup/"
      tag: "Documentation"
      text: "Configure Test Optimization"
    - link: "/tests/explorer/"
      tag: "Documentation"
      text: "Explore Test Results and Performance"
    - link: "/tests/troubleshooting/"
      tag: "Documentation"
      text: "Troubleshooting Test Optimization"
---

Datadog provides official Bazel rules for Test Optimization. Use these rules to configure Bazel test targets. The rules read Test Optimization metadata, write local payload files during test execution, and upload test results to Datadog after Bazel tests complete. See all rules in the [`DataDog/rules_test_optimization` repository](https://github.com/DataDog/rules_test_optimization).

The Bazel integration keeps Datadog metadata fetches outside test execution. During module or repository resolution, Bazel fetches Test Optimization metadata from Datadog and exposes it through a generated repository. During test execution, language-specific macros pass the metadata location to the test process and configure payloads to be written under `TEST_UNDECLARED_OUTPUTS_DIR`. After tests finish, run the doctor and uploader targets with `bazel run`.

> [Test Impact Analysis][1] is not supported for Bazel.

Reviewer 2: ```suggestion
[Test Impact Analysis][1] and automatic [coverage configuration][5] are not supported for Bazel.
```

```text
test:test-optimization --remote_download_outputs=all
```

Without local `test.outputs` directories, the doctor and uploader cannot inspect payload files after `bazel test`.

## Upload payloads

Run the upload flow after your Bazel tests complete. Replace `//tools/test_optimization` with the package where you declared the doctor and uploader targets:

```bash
bazel test --config=test-optimization //...
bazel run --config=test-optimization //tools/test_optimization:dd_test_optimization_doctor
bazel run --config=test-optimization //tools/test_optimization:dd_upload_payloads -- --dry-run --validate-enrichment
DD_API_KEY=<DATADOG_API_KEY> DD_SITE=<DATADOG_SITE> bazel run --config=test-optimization //tools/test_optimization:dd_upload_payloads
```

Do not pass `DD_API_KEY`, `DD_SITE`, `DD_GIT_*`, or upload endpoint variables through `--test_env`. Forward sync metadata with `--repo_env`, and pass upload credentials only to the uploader runtime.

## Further reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /tests/test_impact_analysis/
[2]: /tests/setup/bazel/java/#compatibility
[3]: /tests/setup/bazel/python/#compatibility

> [4]: /tests/setup/bazel/go/#compatibility

Reviewer 2: ```suggestion
[4]: /tests/setup/bazel/go/#compatibility
[5]: /code_coverage/configuration/
```
